### PR TITLE
Add pull request template

### PR DESCRIPTION
**Pull request by johnnovak**
I think it's time to start using a PR template for both internal and external contributions.

On one hand, this sets clear expectations and doesn't make the contributor feel we're "picking on them" when asking for changes, additional testing details, or perhaps a complete re-work. One way to look at the template is "contributor training"—it might take a long time to work your way through the template and the checklist for the _first time_, but then on it should be easy sailing. Core members are already doing the majority of these things anyway, so this should not add significantly to their "workload".

Filling out the template correctly should also free up valuable time on the reviewers' side; having to ask for the same things over and over in PRs gets tiresome and time-consuming (this has become increasingly a problem for me lately).

Requiring proof of manual testing and reproducible test cases are important as not all reviewers are familiar with all parts of the codebase and every little quirky DOSBox feature. I dare to say most of us are quite familiar with one or two major areas, but not so much with the rest (certainly the case for me). Therefore, including these test cases is a necessity so anybody from the team can verify the PRs to some degree before merging.

Lastly, the checklist will help ensure consistent quality and that nothing important gets forgotten. I'm personally guilty of missing a few actions from that checklist from time to time—well, I won't be able to make any excuses going forward 😅 

My hope is that with this template in place, we'll be able to focus more on the fun parts of DOSBox development and reduce "review-related stress"—both on the contributors' and the reviewers' part. Plus asking for reproducible test cases should help reduce the likelihood of functional regressions.

The next comment shows what the template looks like in action.

### CODE REVIEW

**Comment by johnnovak** [comment]
# Description

Please include a clear and concise summary of the changes. Please also include relevant motivation and context. If you're not familiar with the [XY problem](https://xyproblem.info/), please read the linked article first 😄

Feel free to include additional details and explanations if you think it will help the reviewers, but please respect their time and keep it as brief as you can.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that breaks backward compatibility)
- [ ] This change requires a documentation update

# Manual testing

Please describe the tests that you ran to verify your changes.

Provide clear instructions so the reviewers can reproduce and verify your results.

Please also list any relevant details for your test configuration, operating system, etc.

# Checklist

- [ ] I agree to follow the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [ ] My code follows the project's [style guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#coding-style).
- [ ] My commit messages follow the [project's guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] I have performed a self-review of my code.
- [ ] I have commented on the particularly hard-to-understand areas of my code.
- [ ] I split my work into small, bisectable commits, and each commit has a well-defined scope and accurate description.
- [ ] I have [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] I have confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] I have added unit tests where applicable to prove my code works correctly and to avoid future regressions.
- [ ] I have made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] I have [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my changes to the website or the documentation.
- [ ] I have [verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/verify-markdown.sh) that my Markdown file changes pass the linter checks.

File: .github/PULL_REQUEST_TEMPLATE/pull_request_template.md
Comment (kklobe) [inline .github/PULL_REQUEST_TEMPLATE/pull_request_template.md:15]: Do we want a distinct type of "clean up/non-functional" option e.g. my migration of SDL threads to C++ for Voodoo?

Comment (johnnovak) [reply .github/PULL_REQUEST_TEMPLATE/pull_request_template.md:15]: Yes, good idea. Adding it.

Comment (johnnovak) [reply .github/PULL_REQUEST_TEMPLATE/pull_request_template.md:16]: Done.

**Comment by Torinde** [comment]
Fixes #2416

**Comment by kcgen** [comment]
Top level comment is to relax and make the wording a bit more gentle and shorter, as these kinds of things can become onerous and a mental burden (especially for ourselves). I would also drop the checkboxes. 

The visual checklist is a good balance between reminding ourselves but also a heads up of what's expected, but without mandating written attestations for each line item (which can quickly become a burden and scare away small contributions). 

---

Comments:

Type of change section -> suggest replacing this in favour of a single checklist item:
 - applied label(s)

The labels let us rapidly group PRs in searches, where as text strings in the body are harder to see and deal with.

---

Suggest making the checklist start with "I have:", and then each bullet be the action (start with the verb).

**Comment by johnnovak** [comment]
> I think the list of reminders should be a big help to cut down our review workload, which becomes a source of burnout for primary maintainers (which is a sadly recurring theme for open source projects).
> So very happy you've put this together @johnnovak !

Yep, we're in agreement on the high-level goals and the purpose of the template here 😄

> Top level comment is to make the wording a bit more gentle and compact.

Yep, made it more compact, that's good feedback. Well, I think it was "gentle" enough already; I've even trimmed down the "please" words as every second sentence started with "please" 😄 

> Type of change section -> suggest replacing this in favour of a single checklist item:
> The labels let us rapidly group PRs in searches, where as text strings in the body are harder to see and deal with.

Another good feedback, done.

> Suggest making the checklist start with "I have:", and then each bullet be the action (start with the verb).

Good suggestion, done.

> Suggest dropping the requirement to add check boxes or written acknowledgment that the submitter has read and agreed to the checklist. (If DOSBox Staging was a data entry company with pay-per-submission contractors, then yeah: written attestations are warranted! 📋 👮)

On this one, I disagree 😄 We use such checklists in our PR template at work, and if you check out articles about GitHub pull request templates or real-world examples in random repos, many (probably most?) of use checklists. The reason being that they're super useful and help achieve the goals of putting such a checklist there in the first place:

- It's far too easy to just skim through a bulleted list, thinking "yeah, yeah, I've done all that"
- The _act of checking_ is important because it _forces_ you to consider each item carefully. That's by design, and I'm experiencing that at work every day—without having to _mark_ items on the list, I would always forget something.
- If you do your work in multiple sessions (like most people), it's far too easy to forget which points have you already addressed in your previous sessions, and which are still outstanding. The checklist helps keep track of that.

Overall, I'm quite convinced that simply turning the checklist into bullet points would diminish its usefulness greatly. We _want_ people go through the act of _clicking_ the items, really 😄 And checking a few boxes would not slow down experienced team members at all; how long does it take to click 11 times? If the _work_ associated with each checklist item is `N`, then the work plus checking the box would take `N * 1.001` 😆 

> I don't want to overburden ourselves or scare aware "quick and simple" PRs too (_"Awe man, I don't wanna fill out all that paperwork!"_ 🥲 )

Absolutely, me neither! It's about hitting the sweet-spot and making sure the template accomplishes its purpose—otherwise, why bother, right?

But, like I said, for new people you can consider the checklist as "contributor training". Yes, it will take them longer to do things properly for the first few times, but that's _absolutely by design!_ Think about it from this angle:

- New person raises a PR, a "simple" PR
- When I take a look at it, it turns out 80% of the work is still outstanding (e.g. I don't know how to test, commit messages are all over the place, maybe there are broken commits, etc.)
- So at this point, I need to make a choice:
  - Teach the person how to do things properly. I'm always willing to help and explain things, but I really don't have time to go through the basics with every single person. The checklist and the links contained therein take care of 90% of the basic stuff—now the "checklist" has automated this for us. Ta-da! 😎 🎉 
  - Instead of teaching the person, I can help them to finish the remaining 80%, or simply take over. In this case, the contributor _created a task_ for myself! Whether this is help or not is debatable, but the thing is, I'm short of time and I don't want people to create _extra_ tasks for me—I have more than enough already 😄

Ultimately, this is like the Law of Conservation of Energy: if a PR is not fully ready, finishing the job has to be done either by the contributor, or the reviewer/maintainer. One of the main purposes of the PR template/checklist is to consciously push as much of this work to the contributor as we can, because like you said, project maintainers are in short supply, and so is their time.

This does not mean I'm not willing to help people, and you've been very helpful during my first few PRs @kcgen. But we need to be practical here: I agree we should not "scare away" newcomers, but if the checklist filters out people who are unable or unwilling to create PRs that are up to our standards, I'd say "good riddance!" That may sound harsh, but again, reality check: if it's a choice of burning maintainers out by forcing them to finish/babysit unfinished PRs, or filtering out people who are not 100% up for the job, which one would you prefer? (I think I know the answer 😄 ) Again, it's again the Law of Conservation of Energy—_someone_ will need to work more when it comes to unfinished/not-up-to-standard PRs, but who will it be? 😄 

> Will try to add more specifics in a follow up comment.

As I mentioned, many of your comments were good and I've addressed them and I've trimmed down the wording quite a bit—but I feel very strongly about the checkbox. See the current version in my next comment, I consider it pretty much perfect and as slim as it can be. If you have further comments, please also provide change recommendations as I think I've arrived to my personal sweet-spot 😄

**Comment by johnnovak** [comment]
# Description

_Describe a summary of your changes clearly and concisely, including motivation and context._
_Breaking changes need extra explanation on backward compatibility considerations._

_Feel free to include additional details, but please respect the reviewer's time and keep it brief._


## Related issues

_Related issues can be listed here (remove the section if not applicable.)_


# Manual testing

_Please describe the tests that you ran to verify your changes._

_Provide clear instructions so the reviewers can reproduce and verify your results._

_Include relevant details for your test configuration, operating system, etc._


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [ ] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [ ] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)_
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.
- [ ] [verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/verify-markdown.sh) that my website and documentation changes pass the Markdown linter checks.

**Comment by kcgen** [comment]
Re: checkboxes -- oh, I didn't realize the template became somewhat of an interactive form that the submitter can just click to tickoff the items one by one. 

I have only manually created markdown tickboxes (using some kind of `- [x]` notation). 

So I was (hopefully wrongly 😅) assuming the submitter would have to painstakingly creating these markdown lists by hand, instead of simply clicking the template items one by one. 

In that case: fully agree that the sweet spot is achieved 🍬 👍!

**Comment by johnnovak** [comment]
> Re: checkboxes -- oh, I didn't realize the template becomes somewhat of an interactive form that the submitter can just click to tickoff the items one by one.

Yeah, nice isn't it? And super easy and efficient to blast through them! (but maybe don't blast... we want people to consider each one in detail first 😅)
 
> I have only manually created markdown tickboxes (using some kind of `- [x]` notation).

Indeed, we used such a "scheme" at work first, and that was a pain. Clicky-clicky is much nicer 👍🏻 

> In that case: fully agree that the sweet spot is achieved 🍬 👍!

Great 😎 About the rest and my "Law of Conservation of Energy" points; I don't wanna sound like an asshole but if it's a question of pushing 30% more to the contributors versus to the maintainers, I think this is a no-brainer. Pushing all that extra work to the maintainers will just make them quit because _it adds up_—we want to "spread the love", and for casual contributors that's not too much extra effort. Like when we're cleaning the house with my wife—everybody needs to do their share 😄

**Comment by kcgen** [comment]
The more we can punt to CI, the better (even few manual steps for everyone involved):
- I (think?) the markdown linter scans the docs, so maybe it's OK to snip that from the list?
- Although our of scope from this PR, we should try to add the build-commits script to CI toon (and then it's one less thing on this manual checklist). The M1 mini is quite a beast at chewing through the builds, so could be added there.
- In the part where you mention manual testing, could mention using a sanitizer build w/ link to the docs: https://github.com/dosbox-staging/dosbox-staging/blob/main/BUILD.md#make-a-sanitizer-build

**Review by kcgen — APPROVED** [review]
Just a couple remaining minor suggestions. Thanks for putting this together @johnnovak !

Happy to get this change in and everyone using it. 🚀

**Comment by johnnovak** [comment]
>     * I (think?) the markdown linter scans the docs, so maybe it's OK to snip that from the list?

I was considering that, and the script is actually mentioned in the documentation guidelines, so the checklist item is redundant. Removed.

>     * Although out of scope from this PR, we should try to add the build-commits script to CI, too (and then it's one less thing on this manual checklist). The M1 mini is quite a beast at chewing through the builds, so could be added there.

Yeah, that's a good idea, but job for later 😄 

> In the part where you mention manual testing, could mention using a sanitizer build w/ link to the docs: https://github.com/dosbox-staging/dosbox-staging/blob/main/BUILD.md#make-a-sanitizer-build

Mentioned & updated the preview above, you might wanna check the wording.

**Comment by johnnovak** [comment]
> Just a couple remaining minor suggestions. Thanks for putting this together @johnnovak !

Believe it or not, it was suggested by my wife. I kept bitching to her that some contributors are giving me "homework", then she was like "why not make the expectations clear and be done with it?" So that's what I've done 😄

**Comment by johnnovak** [comment]
Merging this now @kcgen ; we can always make enhancements to it later 🥳

**Comment by kcgen** [comment]
Interesting, only after merging the linter has now flagged these in the new template:

- `.github/pull_request_template.md:14` MD025 Multiple top level headers in the same document
- `.github/pull_request_template.md:25` MD025 Multiple top level headers in the same document

MD025: https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md025

(Maybe the linter isn't scanning this directory - or somehow only scan files part of the `main` instead of a branch?)

**Comment by johnnovak** [comment]
> Interesting, only after merging the linter has now flagged these in the new template:
> 
>     * `.github/pull_request_template.md:14` MD025 Multiple top level headers in the same document
> 
>     * `.github/pull_request_template.md:25` MD025 Multiple top level headers in the same document
> 
> 
> MD025: https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md025
> 
> (Maybe the linter isn't scanning this directory - or only scans files part of the `main` instead of a branch?)

I've excluded the Markdown linter from the `.githib` directory, so maybe something else went wrong there. The exclusion worked correctly in this PR and locally.

https://github.com/dosbox-staging/dosbox-staging/pull/2845/commits/e525e452358e4044679dc229c3deafa4f06b91f9

**Pull request merged**
